nerfstudio/engine/trainer.py

```diff
@@ -291,12 +301,22 @@ def train(self) -> None:
 
                 writer.write_out_storage()
 
-        # save checkpoint at the end of training
-        self.save_checkpoint(step)
+        # save checkpoint at the end of training, and write out any remaining events
+        self._after_train()
 
+    def shutdown(self) -> None:
+        """Stop the trainer and stop all associated threads/processes (such as the viewer)."""
+        self.stop_training = True  # tell the training loop to stop
+        if self.viewer_state is not None:
+            self._stop_viewer_server()
```

Comment:
perhaps we could get rid of `_stop_viewer_server()` and just write `self.viewer_state.viser_server.stop()` here?